src/interfaces/permissions-revoke.ts

```diff
@@ -0,0 +1,61 @@
+import type { SignatureInput } from '../types/jws-types.js';
+import type { PermissionsGrantMessage, PermissionsRevokeDescriptor, PermissionsRevokeMessage } from '../types/permissions-types.js';
+
+import { getCurrentTimeInHighPrecision } from '../utils/time.js';
+import { validateAuthorizationIntegrity } from '../core/auth.js';
+import { DwnError, DwnErrorCode } from '../core/dwn-error.js';
+import { DwnInterfaceName, DwnMethodName, Message } from '../core/message.js';
+
+export type PermissionsRevokeOptions = {
+  dateCreated?: string;
+  permissionsGrantId: string;
+  authorizationSignatureInput: SignatureInput;
+};
+
+export class PermissionsRevoke extends Message<PermissionsRevokeMessage> {
+  public static async parse(message: PermissionsRevokeMessage): Promise<PermissionsRevoke> {
+    await validateAuthorizationIntegrity(message);
+
+    return new PermissionsRevoke(message);
+  }
+
+  public static async create(options: PermissionsRevokeOptions): Promise<PermissionsRevoke> {
+    const descriptor: PermissionsRevokeDescriptor = {
+      interface          : DwnInterfaceName.Permissions,
+      method             : DwnMethodName.Revoke,
+      dateCreated        : options.dateCreated ?? getCurrentTimeInHighPrecision(),
+      permissionsGrantId : options.permissionsGrantId,
+    };
+
+    const authorization = await Message.signAsAuthorization(descriptor, options.authorizationSignatureInput);
+    const message: PermissionsRevokeMessage = { descriptor, authorization };
+
+    Message.validateJsonSchema(message);
+
+    return new PermissionsRevoke(message);
+  }
+
+  public async authorize(permissionsGrantMessage: PermissionsGrantMessage): Promise<void> {
+    if (this.author !== permissionsGrantMessage.descriptor.grantedFor) {
+      // Until delegation is implemented, only the DWN owner may grant or revoke access to their DWN
+      throw new DwnError(DwnErrorCode.PermissionsRevokeUnauthorizedRevoke, 'Only the DWN owner may revoke a grant');
+    }
+  }
+
+  public static getOldestRevoke(permissionsRevokeMessages: PermissionsRevokeMessage[]): PermissionsRevokeMessage | undefined {
+    // Sort revoke messages by `dateCreated`
+    const sortedRevokes = permissionsRevokeMessages.sort((revoke1, revoke2) => {
```

Comment:
Nit: O(nLog(n)) can be avoid with an O(n) here. Basically like how `getNewestMessage()` is done.

I know #406 is probably putting a lot uncertainty here, so I'd be fine if you just reference #406 with a TODO comment to kick the can, 'coz we could reuse much of the existing code.

Reply:
Good call. Fixed.